client/order/orderbook.go

```diff
@@ -0,0 +1,226 @@
+package order
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/decred/dcrdex/dex/msgjson"
+)
+
+// RemoteOrderBook defines the functions a client tracked order book
+// must implement.
+type RemoteOrderBook interface {
+	// Initialize instantiates a client tracked order book with the
+	// current order book snapshot.
+	Initialize(*msgjson.OrderBook)
+	// Book adds a new order to the order book.
+	Book(*msgjson.BookOrderNote)
+	// Unbook removes an order from the order book.
+	Unbook(*msgjson.UnbookOrderNote) error
+}
+
+// OrderBook represents an client tracked order book.
+type OrderBook struct {
+	seq      uint64
+	marketID string
+	orders   map[string]*msgjson.BookOrderNote
+	buys     map[uint64][]*msgjson.BookOrderNote
+	sells    map[uint64][]*msgjson.BookOrderNote
+	synced   bool
+	mtx      sync.RWMutex
+}
+
+// addOrder adds an order to the order book. This must be called with
+// the order book mutex held.
+func (ob *OrderBook) addOrder(order *msgjson.BookOrderNote) {
+	ob.orders[order.OrderID.String()] = order
+
+	switch order.Side {
+	case msgjson.BuyOrderNum:
+		_, ok := ob.buys[order.Rate]
+		if !ok {
+			ob.buys[order.Rate] = make([]*msgjson.BookOrderNote, 0)
+		}
+		ob.buys[order.Rate] = append(ob.buys[order.Rate], order)
+
+	case msgjson.SellOrderNum:
+		_, ok := ob.sells[order.Rate]
+		if !ok {
+			ob.sells[order.Rate] = make([]*msgjson.BookOrderNote, 0)
+		}
+		ob.sells[order.Rate] = append(ob.sells[order.Rate], order)
+	}
+}
+
+// removeOrder removes an order from the order book. This must be called with
+// the order book mutex held.
+func (ob *OrderBook) removeOrder(order *msgjson.BookOrderNote) error {
+	switch order.Side {
+	case msgjson.BuyOrderNum:
+		ordersAtPrice, ok := ob.buys[order.Rate]
+		if !ok {
+			return fmt.Errorf("no buy orders found at price %v", order.Rate)
+		}
+
+		for idx := len(ordersAtPrice) - 1; idx >= 0; idx-- {
+			if bytes.Equal(ordersAtPrice[idx].OrderID, order.OrderID) {
+				ordersAtPrice = append(ordersAtPrice[:idx], ordersAtPrice[idx+1:]...)
+				ob.buys[order.Rate] = ordersAtPrice
+				break
+			}
+		}
+
+	case msgjson.SellOrderNum:
+		ordersAtPrice, ok := ob.sells[order.Rate]
+		if !ok {
+			return fmt.Errorf("no sell orders found at price %v", order.Rate)
+		}
+
+		for idx := len(ordersAtPrice) - 1; idx >= 0; idx-- {
+			if bytes.Equal(ordersAtPrice[idx].OrderID, order.OrderID) {
+				ordersAtPrice = append(ordersAtPrice[:idx], ordersAtPrice[idx+1:]...)
+				ob.sells[order.Rate] = ordersAtPrice
+				break
+			}
+		}
+	}
+
+	delete(ob.orders, order.OrderID.String())
```

Comment:
Seems like having `buyOrders` and `sellOrders` would be more efficient.

Reply:
On the other hand, for `UnbookOrderNote`, where you won't have the side as part of the notification, it's better to have a single order array here.

Reply:
Checking one and then the other wouldn't be bad, but that's true.